### DISCUSSION TITLE
How to check feasibility, i.e. whether a given point for x satisfies the constraints

### DISCUSSION THREAD

**Discussion by FabianHuebenthal** [casadi/casadi#3383] [Q&A]
Hi CasADi-team,

Given a constrained optimization problem formulated using CasADi (Python), either NLP interface or Opti-Stack interface, how can one check the feasibility of certain values of the optimization variables? Is there a method like opti.is_feasible(x_to_check)?

Below is a minimal working example and my questions as comments at the appropriate positions.

Thanks in advance!

PS: And I hope you can excuse it if the question has already been answered elsewhere, but I couldn't find a suitable question so far.

```python
# external imports
from casadi import *

# create opti stack object
opti = casadi.Opti()

# define optimization variables
x = opti.variable(2)

# define parameters
p = opti.parameter(4)
opti.set_value(p, [1, 1, 1, 1])

# define objective function
f = (p[1]*x[1] - p[0]*x[0]**2)**2
opti.minimize(f)

# inequality constraints
g = -x[0] - x[1] <= -1
opti.subject_to(g)

# equality constraints
h = p[2]*x[0]**2 + p[3]*x[1]**2 == 1
opti.subject_to(h)

# select solver
############################################
# Question
############################################
# What is the key difference between p_opts and s_opts? Is my current understanding correct?
# p_opts = generic solver settings?
# s_opts = solver specific settings?
############################################
p_opts = {
    'expand':True,
    'print_time':0,  # non-ipopt-specific setting to suppress print information about execution time
    'ipopt.print_level':0  # ipopt-specific setting to suppress ipopt print messages
    }  # CasADi plugin options, generic options
s_opts = {
    'max_iter': 100
    }  # optional solver-specific options
opti.set_initial(x, [0, 0])  # set initial guess of the solution
opti.solver('ipopt', p_opts, s_opts)

# solve NLP
sol = opti.solve()

# print status information
print(opti.stats())
print(opti.return_status())

############################################
# Question
############################################
# check feasibility of any point x
x_to_check = [0, 1]
opti.is_feasible(x_to_check)  # Does a method exist to check feasibility?
############################################

# print solution
print("-----")
print("objective at solution = ", sol.value(f))
print("primal solution = ", sol.value(x))
print("dual solution (g) = ", sol.value(opti.lam_g))
```

**Comment by MHW-UDS** [comment]
If you want to check if a pair (x',p') is a feasible (but not necessarily optimal) solution to the optimization problem min_x f(x,p) s.t. x_lb<= x<=x_ub, g_lb<=g(x,p)<=g_ub you can simpliy check if all the inequalities are satisfied (or sufficiently close to being satisfied, e.g. g_fun(x',p')-g_ub<tol). You can create the function g_fun by casadi.Function('g',{x,p},{g}) (Matlab syntax). Since I dont use opti I don't know if you can obtain x,p,g by opti.x, opti.p, opti.g or differently.

If you want to check for optimality look at the KKT conditions for Casadis problem formulation in eq. (8) of the paper:
Andersson, Joel A.E., und James B. Rawlings. „Sensitivity Analysis for Nonlinear Programming in CasADi“. In IFAC-PapersOnLine, 51:331–36, 2018. https://doi.org/10.1016/j.ifacol.2018.11.055.
The diagonals entries of the matrices in (8d), (8e) can be obtained via 
```
function [I_y_v,I_y_lb_v,I_y_ub_v] = I_active_set_lam(lam_y,tol)
    lam_y = full(lam_y);

    I_y_lb_v = double(lam_y<-tol);
    I_y_ub_v = double(lam_y> tol); 

    I_y_v = I_y_lb_v+I_y_ub_v;
end
```
(Matlab syntax) where you can plug in lam_x and lam_g (tol should be smaller than 10^-24).
